django_fido/backends.py

```diff
@@ -56,24 +85,36 @@ def authenticate(self, request: HttpRequest, user: AbstractBaseUser, fido2_serve
             raise PermissionDenied("Counter didn't increase.")
         return user
 
-    def mark_device_used(self, device, counter):
-        """Update FIDO 2 device usage information."""
-        if counter == 0 and device.counter == 0:
-            # Counter is unsupported by the device, bail out early
-            return
-        if counter <= device.counter:
-            _LOGGER.info("FIDO 2 authentication failed because of not increasing counter.")
-            raise ValueError("Counter didn't increase.")
-        device.counter = counter
-        device.full_clean()
-        device.save()
 
-    def get_user(self, user_id):
-        """Return user based on its ID."""
+class Fido2PasswordlessAuthenticationBackend(BaseFido2AuthenticationBackend):
+    """Authenticate user using FIDO 2 passwordlessly using supplied user handle."""
+
+    def authenticate(self, request: HttpRequest, user: Optional[AbstractBaseUser], fido2_server: Fido2Server,
+                     fido2_state: Dict[str, bytes], fido2_response: Dict[str, Any]) -> Optional[AbstractBaseUser]:
+        """Authenticate using FIDO 2."""
+        user_handle = fido2_response['user_handle']
         try:
-            return get_user_model().objects.get(pk=user_id)
-        except get_user_model().DoesNotExist:
+            authenticator = Authenticator.objects.get(user_handle=user_handle)
+            user = authenticator.user
+            credentials = [authenticator.credential]
+            credential = fido2_server.authenticate_complete(
+                fido2_state, credentials, fido2_response['credential_id'], fido2_response['client_data'],
+                fido2_response['authenticator_data'], fido2_response['signature'])
+        except ValueError as error:
+            _LOGGER.info("FIDO 2 authentication failed with error: %r", error)
             return None
+        except Authenticator.DoesNotExist:
+            _LOGGER.info("FIDO 2 authentication could not find user handle: %s", user_handle)
+            return None
+
+        device = user.authenticators.get(credential_id_data=base64.b64encode(credential.credential_id).decode('utf-8'))
```

Comment:
I believe that `device == authenticator` so this query is not needed.

Also I'd rename `authenticator` variable to `device` to keep naming conventions within the file.

Reply:
Hi sorry I will get to this soon.

Reply:
I have made most of the changes, but I cannot find `device == authenticator` in my code

Reply:
The code is not there, I just wanted to emphasize, that they are the same object. Sorry for the confusion :blush: 

Now, the removal of the second query should be enough.
```suggestion
```

Reply:
Sweet, done.